includes/class-wc-payments-token-service.php

```diff
@@ -213,6 +205,77 @@ public function woocommerce_get_customer_payment_tokens( $tokens, $user_id, $gat
 		return $tokens;
 	}
 
+	/**
+	 * Retrieves the payment method types for which tokens should be retrieved.
+	 *
+	 * This function determines the appropriate payment method types based on the provided gateway ID.
+	 * - If a gateway ID is provided, it retrieves the payment methods specific to that gateway to prevent duplication of saved tokens under incorrect payment methods during checkout.
+	 * - If no gateway ID is provided, it retrieves the default payment methods to fetch all saved tokens, e.g., for the Blocks checkout or My Account page.
+	 *
+	 * @param string|null $gateway_id The optional ID of the gateway.
+	 * @return array The list of retrievable payment method types.
+	 */
+	private function get_retrievable_payment_method_types( $gateway_id = null ) {
+		if ( empty( $gateway_id ) ) {
+			return $this->get_all_retrievable_payment_types();
+		} else {
+			return $this->get_gateway_specific_retrievable_payment_types( $gateway_id );
+		}
+	}
+
+	/**
+	 * Returns all the enabled retrievable payment method types.
+	 *
+	 * @return array Enabled retrievable payment method types.
+	 */
+	private function get_all_retrievable_payment_types() {
+		$types = [ Payment_Method::CARD ];
+
+		if ( $this->is_payment_method_enabled( Payment_Method::SEPA ) ) {
+			$types[] = Payment_Method::SEPA;
+		}
+
+		if ( $this->is_payment_method_enabled( Payment_Method::LINK ) ) {
+			$types[] = Payment_Method::LINK;
+		}
+
+		return $types;
+	}
+	/**
+	 * Returns retrievable payment method types for a given gateway.
+	 *
+	 * @param string $gateway_id The ID of the gateway.
+	 * @return array Retrievable payment method types for the specified gateway.
+	 */
+	private function get_gateway_specific_retrievable_payment_types( $gateway_id ) {
+		$types = [];
+
+		foreach ( self::REUSABLE_GATEWAYS_BY_PAYMENT_METHOD as $payment_method => $gateway ) {
+			if ( $gateway === $gateway_id ) {
+				// Stripe Link is part of the card gateway, so we need to check separately if Link is enabled.
+				if ( Payment_Method::LINK === $payment_method ) {
+					if ( $this->is_payment_method_enabled( Payment_Method::LINK ) ) {
+						$types[] = Payment_Method::LINK;
+					}
+				} else {
+					$types[] = $payment_method;
+				}
+			}
+		}
```

Comment:
```suggestion
		foreach ( self::REUSABLE_GATEWAYS_BY_PAYMENT_METHOD as $payment_method => $gateway ) {
			if ( $gateway !== $gateway_id ) {
				continue;
			}
			
			// Stripe Link is part of the card gateway, so we need to check separately if Link is enabled.
			if ( Payment_Method::LINK === $payment_method && ! $this->is_payment_method_enabled( Payment_Method::LINK ) ) {
				continue;
			}
			
			$types[] = $payment_method;
		}
```
It's not a huge deal and probably a matter of personal preference, but we could avoid multiple layers of nested conditionals here by rewriting this block as something similar to the above. I think this reads a little cleaner, since it more explicitly conveys the conditions under which certain payment method types will be excluded and there is less branching logic to follow, which can be a little confusing for small brains like mine. 😅 

Let me know what you think: in either case, doesn't really affect performance, so up to you to include or ignore. 🙏

Reply:
I like the suggestion, it's easier to understand, thank you! ece0e83715a7e40cdc4d58b2bec198ba2e626d32
I also went through all the testing instructions once again to confirm that everything works 👍